x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java

```diff
@@ -328,14 +328,19 @@ static DataStream updateLocalDataStream(Index backingIndexToFollow, DataStream l
                 );
             }
 
-            List<Index> backingIndices = new ArrayList<>(localDataStream.getIndices());
-            backingIndices.add(backingIndexToFollow);
-
-            // When following an older backing index it should be positioned before the newer backing indices.
-            // Currently the assumption is that the newest index (highest generation) is the write index.
-            // (just appending an older backing index to the list of backing indices would break that assumption)
-            // (string sorting works because of the naming backing index naming scheme)
-            backingIndices.sort(Comparator.comparing(Index::getName));
+            final List<Index> backingIndices;
+            if (localDataStream.getIndices().contains(backingIndexToFollow) == false) {
```

Comment:
I would find it more explicit to use the `currentState` above to detect that the index already existed and thus should not be added to the data stream? (and perhaps rename `currentState` to `previousState` for readability).

We should then assert that it already exists there. And here we can assert that it cannot exist already (or rely on the new assertion in the `DataStream` constructor).

Reply:
I'd rather not use the `currentState` here to be honest. It will work now, but we go through super complicated code when restoring the datastream in the restore service, I find it much easier to reason about taking the information from the builder.
If I use the `currentState` then the whole logic becomes predicated on a bunch of assumptions (like that we won't ever remove an index in the restore action) and as a result (at least to me) much less obviously correct.

Reply:
I'd hope we can find a better way to signal this then, perhaps expand the updater contract to include this boolean information. I am OK with merging as is, but would like to see this in a follow-up.

The current way seems to be somewhat lenient and could "repair" something inadvertently.